### PR TITLE
Nexus: Extend excitations to allow 'lowest' gap

### PR DESCRIPTION
**Pull request by mcbennet**
## Proposed changes

This PR allows the user to specify 'lowest' for the excitation, e.g., 

```python3
generate_qmcpack(
...,
excitation = ['up','lowest'], # 'up' can be replaced with 'down', 'singlet', or 'triplet'
...,
)
```
where nexus will then generate the lowest excitation for the given type (up,down,singlet,triplet) based on the energy index (electron count). The resulting qmcpack input will contain the corresponding excitation in energy format.

This PR also fixes a couple minor bugs in the previous 'singlet' and 'triplet' types.

In addition, post analysis checks for the various excitation input types have been added -- the user's input is checked against qmcpack's generated einspline files to verify that the user's requested excitation is consistent with the change to the einspline file. 

## What type(s) of changes does this code introduce?

- Bugfix
- New feature

### Does this introduce a breaking change?

- No. nxs-tests pass in ubuntu desktop (with the exception of `simulation_module`)

## What systems has this change been tested on?

Ubuntu desktop

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- N/A. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by mcbennet** [comment]
Changing to WIP. There are a few additional tests I would like to look at before this goes in.

**Comment by mcbennet** [comment]
Okay, IMO this PR is ready to go in -- removing draft state. For diamond (primitive with 4 up and 4 down electrons), I tested that the VMC-NoJ kinetic energies from various excitation (gamma-to-gamma) specifications agree with the total kinetic energy from the `eshdf` tool. That plot is attached.

**Ground**:      ground state WF
**Up-Band**:    excitation = ['up','0 3 0 4']
**Dn-Band**:    excitation = ['down','0 3 0 4']
**Up-Energy**: excitation = ['up','-4 +5']
**Up-Binary**:   excitation = ['singlet','-4 +5'] -> Before running, CSF coefficients are changed by hand to make excitation consistent with others above, i.e., (0.707,0.707) -> (1.000,0.000)

All the VMC KEs appear to be in really good agreement here and also in good agreement with `eshdf`.

Screenshot from 2021-11-24 08-28-37

**Comment by jtkrogel** [comment]
Looking good overall.  Just needs some hardening.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]
I'm in favor of merging this PR. On the other hand, I'm having difficulty to understand these input strings due to the lack of documentation neither in qmcpack nor in nexus.

**Comment by prckent** [comment]
Are there any relevant docs that could be updated? The "documentation gap" around NEXUS and different ways of driving QMC is growing. This change has the wrong sign.

**Comment by mcbennet** [comment]
I had a discussion with Jaron about this. I will include this as an alternative in the VMC Bulk diamond example in the documentation. Also, I will supply an additional nexus example that uses this functionality.

**Comment by mcbennet** [comment]
This PR should now be ready for review.

**Comment by ye-luo** [comment]
Test this please

**Review by jtkrogel — CHANGES_REQUESTED** [review]
See comments above.

**Comment by mcbennet** [comment]
@jtkrogel I think everything in your review is either addressed or outdated

File: nexus/examples/qmcpack/rsqmc_misc/excited/vmc_excitation_alternatives.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+            )
+        ],
+    dependencies = [(conv,'orbitals'),
+                    (opt,'jastrow')],
+    )
+
+################################################################################
+############ Single Determinant Excitations ####################################
+################################################################################
+
```
Comment (jtkrogel) [inline nexus/examples/qmcpack/rsqmc_misc/excited/vmc_excitation_alternatives.py:137]: Make a note that all the below do the exact same thing and just use alternative syntax.

File: nexus/lib/qmcpack.py [hunk 2]
```diff
@@ -509,6 +510,220 @@ def post_analyze(self,analyzer):
...
+                        else:
+                            sdet =  wf.determinantset.get('downdet')
+                        #end if
+                        from numpy import linalg,where,isclose
+                        vb = int(sdet.size / abs(linalg.det(tilematrix))) -1  # Separate for each spin channel
+                        cb = vb+1
+                        # Convert band_1, band_2 to band indexes
+                        bands = [band_1, band_2]
+                        for bnum, b in enumerate(bands):
+                            if 'CB' in b:
```
[the remark this answers is not in the document]
Comment (jtkrogel) [reply nexus/lib/qmcpack.py:634]: Still needed

File: nexus/lib/qmcpack.py [hunk 3]
```diff
@@ -509,6 +511,224 @@ def post_analyze(self,analyzer):
...
+                        orb2 = orb1+1 
+                    else:
+                        orb1 = int(exc_input[1].split()[0][1:])
+                        orb2 = int(exc_input[1].split()[1][1:])
+                    #end if
+                    if exc1 in ('up','down'):
+
+                        spin_channel = exc1
+                        nelec = elns.groups[spin_channel[0]].size
+
```
Comment (jtkrogel) [inline nexus/lib/qmcpack.py:574]: eigs_spin = edata[spin_channel].Energy
(and reuse below)

File: nexus/lib/qmcpack.py [hunk 4]
```diff
@@ -509,6 +511,224 @@ def post_analyze(self,analyzer):
...
+                        orb1_eig = sorted(edata[spin_channel].Energy)[orb1-1]
```
Comment (jtkrogel) [inline nexus/lib/qmcpack.py]: sort once, then read twice

File: nexus/lib/qmcpack.py [hunk 5]
```diff
@@ -509,6 +511,224 @@ def post_analyze(self,analyzer):
...
+                        orb2_eig = sorted(edata[spin_channel].Energy)[orb2-1]
+
+                        if orb1_eig not in edata[spin_channel]['Energy'][nelec:]:
```
Comment (jtkrogel) [inline nexus/lib/qmcpack.py:579]: Possible precision problem on floating point comparison.
Likely something like the below is better:

etol = 1e-6
if np.abs(eigs_spin[nelec:]-orb1_eig).max() > tol:

File: nexus/lib/qmcpack.py [hunk 6]
```diff
@@ -509,6 +510,220 @@ def post_analyze(self,analyzer):
...
+                                if '-' in b:
+                                    b = b.split('-')
+                                    bands[bnum] = cb - int(b[1])
+                                elif '+' in b:
+                                    b = b.split('+')
+                                    bands[bnum] = cb + int(b[1])
+                                else:
+                                    bands[bnum] = cb
+                                #end if
+                            elif 'VB' in b:
```
[the remark this answers is not in the document]
Comment (jtkrogel) [reply nexus/lib/qmcpack.py:645]: Still needed

File: nexus/lib/qmcpack.py [hunk 7]
```diff
@@ -509,6 +511,224 @@ def post_analyze(self,analyzer):
...
+                        kpath_label = array(kpath['explicit_kpoints_labels'])
+                        kpath_rel   = kpath['explicit_kpoints_rel']
+                        
+                        k1_in = k_1
+                        k2_in = k_2
+                        if k_1 in kpath_label and k_2 in kpath_label:   
+                            k_1 = kpath_rel[where(kpath_label == k_1)][0]
+                            k_2 = kpath_rel[where(kpath_label == k_2)][0]
+
+                            #kpts = nscf.input.k_points.kpoints
```
Comment (jtkrogel) [inline nexus/lib/qmcpack.py:676]: Delete

File: nexus/lib/qmcpack_input.py [hunk 8]
```diff
@@ -4727,6 +4734,104 @@ def generate_determinantset(up             = 'u',
...
+        msg += 'or lowest format (e.g. "lowest").\n'
+        msg += 'You Provided: {0}'
+        msg = msg.format(excitation)
+
+        QmcpackInput.class_error(msg)
+
+    #end if
+
+    return exc_spin,exc_type,exc_spins,exc_types,exc1,exc2
+
```
Comment (jtkrogel) [inline nexus/lib/qmcpack_input.py:4830]: Delete empty lines following return

File: nexus/lib/qmcpack_input.py [hunk 9]
```diff
@@ -4980,7 +5069,11 @@ def generate_determinantset_old(type           = 'bspline',
             band_1, band_2 = bands
             
             # Convert k_1 k_2 to wavevector indexes
-            structure   = system.structure.folded_structure.copy()
+            if system.structure.has_folded():
```
Comment (jtkrogel) [inline nexus/lib/qmcpack_input.py:5072]: Also use `.get_smallest()` here

**Review by jtkrogel — CHANGES_REQUESTED** [review]

File: nexus/lib/qmcpack.py (hunk 5 above)
Comment (mcbennet) [reply nexus/lib/qmcpack.py]: There was an issue with our solution here if orb1 and orb2 were degenerate. But I think the latest commit resolves this

**Review by jtkrogel — APPROVED** [review]
LGTM

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**